File: node/16.20.1-bookworm-slim-fips/diagnostics/test-connectivity.js
#!/usr/bin/env node
/**
 * Connectivity Tests
 * Tests HTTPS/TLS connectivity with FIPS-approved protocols
 */

const https = require('https');
const tls = require('tls');
const fs = require('fs');

class ConnectivityTests {
    constructor() {
        this.results = {
            test_area: '2-connectivity',
            timestamp: new Date().toISOString(),
            container: 'node:16.20.1-bookworm-slim-fips',
            total_tests: 8,
            passed: 0,
            failed: 0,
            skipped: 0,
            tests: []
        };
    }

    logTest(testId, name, status, details = '', durationMs = 0) {
        const testResult = {
            id: testId,
            name: name,
            status: status,
            duration_ms: durationMs,
            details: details
        };
        this.results.tests.push(testResult);

        const symbols = { pass: '✅', fail: '❌', skip: '⏭️' };
        console.log(`${symbols[status]} ${testId} ${name}: ${status.toUpperCase()}`);
        if (details) {
            console.log(`   Details: ${details}`);
        }
        console.log('');

        if (status === 'pass') this.results.passed++;
        else if (status === 'fail') this.results.failed++;
        else this.results.skipped++;
    }

    async test_2_1_https_get_request() {
        return new Promise((resolve) => {
            try {
                const startTime = Date.now();
                console.log(`  Testing HTTPS GET request to www.google.com...`);

                const options = {
                    hostname: 'www.google.com',
                    port: 443,
                    path: '/',
                    method: 'GET',
                    timeout: 10000
                };

                const req = https.request(options, (res) => {
                    const duration = Date.now() - startTime;

                    console.log(`  Status Code: ${res.statusCode}`);
                    console.log(`  TLS Protocol: ${res.socket.getProtocol()}`);

                    const cipher = res.socket.getCipher();
                    if (cipher) {
                        console.log(`  Cipher: ${cipher.name}`);
                        console.log(`  TLS Version: ${cipher.version}`);
                    }

                    let data = '';
                    res.on('data', (chunk) => { data += chunk; });
                    res.on('end', () => {
                        console.log(`  Response size: ${data.length} bytes`);

                        if (res.statusCode === 200 || res.statusCode === 301 || res.statusCode === 302) {
                            this.logTest('2.1', 'HTTPS GET Request', 'pass',
                                `Connected via ${res.socket.getProtocol()} with ${cipher?.name}`, duration);
                        } else {
                            this.logTest('2.1', 'HTTPS GET Request', 'fail',
                                `Unexpected status code: ${res.statusCode}`, duration);
                        }
                        resolve();
                    });
                });

                req.on('error', (error) => {
                    this.logTest('2.1', 'HTTPS GET Request', 'fail', error.message);
                    resolve();
                });

                req.on('timeout', () => {
                    req.destroy();
                    this.logTest('2.1', 'HTTPS GET Request', 'fail', 'Request timeout');
                    resolve();
                });

                req.end();
            } catch (error) {
                this.logTest('2.1', 'HTTPS GET Request', 'fail', error.message);
                resolve();
            }
        });
    }

    async test_2_2_tls_protocol_support() {
        return new Promise((resolve) => {
            try {
                console.log(`  Testing TLS protocol support...`);

                const options = {
                    host: 'www.google.com',
                    port: 443,
                    servername: 'www.google.com', // SNI support
                    timeout: 10000
                };

                const socket = tls.connect(options, () => {
                    const protocol = socket.getProtocol();
                    const cipher = socket.getCipher();

                    console.log(`  Connected Protocol: ${protocol}`);
                    console.log(`  Cipher Suite: ${cipher.name}`);
                    console.log(`  Cipher Version: ${cipher.version}`);

                    socket.end();

                    // TLS 1.2 or TLS 1.3 are FIPS-approved
                    if (protocol === 'TLSv1.2' || protocol === 'TLSv1.3') {
                        this.logTest('2.2', 'TLS Protocol Support', 'pass',
                            `${protocol} with ${cipher.name}`);
                    } else {
                        this.logTest('2.2', 'TLS Protocol Support', 'fail',
                            `Unexpected protocol: ${protocol}`);
                    }
                    resolve();
                });

                socket.on('error', (error) => {
                    this.logTest('2.2', 'TLS Protocol Support', 'fail', error.message);
                    resolve();
                });

                socket.setTimeout(10000, () => {
                    socket.destroy();
                    this.logTest('2.2', 'TLS Protocol Support', 'fail', 'Connection timeout');
                    resolve();
                });
            } catch (error) {
                this.logTest('2.2', 'TLS Protocol Support', 'fail', error.message);
                resolve();
            }
        });
    }

    async test_2_3_tls_1_2_connection() {
        return new Promise((resolve) => {
            try {
                console.log(`  Testing TLS 1.2 connection...`);

                const options = {
                    host: 'www.google.com',
                    port: 443,
                    servername: 'www.google.com', // SNI support
                    minVersion: 'TLSv1.2',
                    maxVersion: 'TLSv1.2',
                    timeout: 10000
                };

                const socket = tls.connect(options, () => {
                    const protocol = socket.getProtocol();
                    const cipher = socket.getCipher();

                    console.log(`  Protocol: ${protocol}`);
                    console.log(`  Cipher: ${cipher.name}`);

                    socket.end();

                    if (protocol === 'TLSv1.2') {
                        this.logTest('2.3', 'TLS 1.2 Connection', 'pass',
                            `TLS 1.2 connection successful with ${cipher.name}`);
                    } else {
                        this.logTest('2.3', 'TLS 1.2 Connection', 'fail',
                            `Expected TLSv1.2, got ${protocol}`);
                    }
                    resolve();
                });

                socket.on('error', (error) => {
                    this.logTest('2.3', 'TLS 1.2 Connection', 'fail', error.message);
                    resolve();
                });

                socket.setTimeout(10000, () => {
                    socket.destroy();
                    this.logTest('2.3', 'TLS 1.2 Connection', 'fail', 'Connection timeout');
                    resolve();
                });
            } catch (error) {
                this.logTest('2.3', 'TLS 1.2 Connection', 'fail', error.message);
                resolve();
            }
        });
    }

    async test_2_4_tls_1_3_connection() {
        return new Promise((resolve) => {
            try {
                console.log(`  Testing TLS 1.3 connection...`);

                const options = {
                    host: 'www.google.com',
                    port: 443,
                    servername: 'www.google.com', // SNI support
                    minVersion: 'TLSv1.3',
                    maxVersion: 'TLSv1.3',
                    timeout: 10000
                };

                const socket = tls.connect(options, () => {
                    const protocol = socket.getProtocol();
                    const cipher = socket.getCipher();

                    console.log(`  Protocol: ${protocol}`);
                    console.log(`  Cipher: ${cipher.name}`);

                    socket.end();

                    if (protocol === 'TLSv1.3') {
                        this.logTest('2.4', 'TLS 1.3 Connection', 'pass',
                            `TLS 1.3 connection successful with ${cipher.name}`);
                    } else {
                        this.logTest('2.4', 'TLS 1.3 Connection', 'fail',
                            `Expected TLSv1.3, got ${protocol}`);
                    }
                    resolve();
                });

                socket.on('error', (error) => {
                    // TLS 1.3 may not be supported by all servers
                    console.log(`  Note: ${error.message}`);
                    this.logTest('2.4', 'TLS 1.3 Connection', 'skip',
                        'TLS 1.3 not supported by server (acceptable)');
                    resolve();
                });

                socket.setTimeout(10000, () => {
                    socket.destroy();
                    this.logTest('2.4', 'TLS 1.3 Connection', 'fail', 'Connection timeout');
                    resolve();
                });
            } catch (error) {
                this.logTest('2.4', 'TLS 1.3 Connection', 'fail', error.message);
                resolve();
            }
        });
    }

    async test_2_5_certificate_validation() {
        return new Promise((resolve) => {
            try {
                console.log(`  Testing certificate validation...`);

                const options = {
                    host: 'www.google.com',
                    port: 443,
                    servername: 'www.google.com', // SNI support
                    rejectUnauthorized: true, // Enforce certificate validation
                    timeout: 10000
                };

                const socket = tls.connect(options, () => {
                    const cert = socket.getPeerCertificate();

                    console.log(`  Certificate subject: ${cert.subject?.CN || 'N/A'}`);
                    console.log(`  Certificate issuer: ${cert.issuer?.CN || 'N/A'}`);
                    console.log(`  Valid from: ${cert.valid_from}`);
                    console.log(`  Valid to: ${cert.valid_to}`);
                    console.log(`  Authorized: ${socket.authorized}`);

                    socket.end();

                    if (socket.authorized) {
                        this.logTest('2.5', 'Certificate Validation', 'pass',
                            `Certificate validated for ${cert.subject?.CN}`);
                    } else {
                        this.logTest('2.5', 'Certificate Validation', 'fail',
                            `Certificate not authorized: ${socket.authorizationError}`);
                    }
                    resolve();
                });

                socket.on('error', (error) => {
                    this.logTest('2.5', 'Certificate Validation', 'fail', error.message);
                    resolve();
                });

                socket.setTimeout(10000, () => {
                    socket.destroy();
                    this.logTest('2.5', 'Certificate Validation', 'fail', 'Connection timeout');
                    resolve();
                });
            } catch (error) {
                this.logTest('2.5', 'Certificate Validation', 'fail', error.message);
                resolve();
            }
        });
    }

    async test_2_6_cipher_suite_negotiation() {
        return new Promise((resolve) => {
            try {
                console.log(`  Testing cipher suite negotiation...`);

                const options = {
                    host: 'www.google.com',
                    port: 443,
                    servername: 'www.google.com', // SNI support
                    timeout: 10000
                };

                const socket = tls.connect(options, () => {
                    const cipher = socket.getCipher();
                    const protocol = socket.getProtocol();

                    console.log(`  Negotiated cipher: ${cipher.name}`);
                    console.log(`  Protocol: ${protocol}`);
                    console.log(`  Cipher version: ${cipher.version}`);

                    socket.end();

                    // Check for FIPS-approved cipher patterns
                    const cipherName = cipher.name.toLowerCase();
                    const isFipsCompliant =
                        cipherName.includes('aes') &&
                        (cipherName.includes('gcm') || cipherName.includes('ccm'));

                    console.log(`  FIPS-compliant: ${isFipsCompliant}`);

                    if (isFipsCompliant) {
                        this.logTest('2.6', 'Cipher Suite Negotiation', 'pass',
                            `FIPS-approved cipher negotiated: ${cipher.name}`);
                    } else {
                        this.logTest('2.6', 'Cipher Suite Negotiation', 'fail',
                            `Non-FIPS cipher: ${cipher.name}`);
                    }
                    resolve();
                });

                socket.on('error', (error) => {
                    this.logTest('2.6', 'Cipher Suite Negotiation', 'fail', error.message);
                    resolve();
                });

                socket.setTimeout(10000, () => {
                    socket.destroy();
                    this.logTest('2.6', 'Cipher Suite Negotiation', 'fail', 'Connection timeout');
                    resolve();
                });
            } catch (error) {
                this.logTest('2.6', 'Cipher Suite Negotiation', 'fail', error.message);
                resolve();
            }
        });
    }

    async test_2_7_concurrent_connections() {
        try {
            console.log(`  Testing concurrent HTTPS connections...`);

            const hosts = [
                'www.google.com',
                'www.github.com',
                'www.amazon.com'
            ];

            const startTime = Date.now();

            const requests = hosts.map((hostname) => {
                return new Promise((resolve) => {
                    const options = {
                        hostname: hostname,
                        port: 443,
                        path: '/',
                        method: 'HEAD',
                        timeout: 10000
                    };

                    const req = https.request(options, (res) => {
                        const cipher = res.socket.getCipher();
                        console.log(`    ✓ ${hostname}: ${res.statusCode} (${cipher?.name})`);
                        resolve(true);
                    });

                    req.on('error', (error) => {
                        console.log(`    ✗ ${hostname}: ${error.message}`);
                        resolve(false);
                    });

                    req.on('timeout', () => {
                        req.destroy();
                        console.log(`    ✗ ${hostname}: timeout`);
                        resolve(false);
                    });

                    req.end();
                });
            });

            const results = await Promise.all(requests);
            const duration = Date.now() - startTime;
            const successCount = results.filter(r => r).length;

            console.log(`  Successful connections: ${successCount}/${hosts.length}`);
            console.log(`  Total time: ${duration}ms`);

            if (successCount >= 2) {
                this.logTest('2.7', 'Concurrent Connections', 'pass',
                    `${successCount}/${hosts.length} concurrent connections successful`, duration);
            } else {
                this.logTest('2.7', 'Concurrent Connections', 'fail',
                    `Only ${successCount}/${hosts.length} connections succeeded`, duration);
            }
        } catch (error) {
            this.logTest('2.7', 'Concurrent Connections', 'fail', error.message);
        }
    }

    async test_2_8_https_post_request() {
        return new Promise((resolve) => {
            try {
                console.log(`  Testing HTTPS POST request...`);

                const postData = JSON.stringify({
                    test: 'fips-connectivity',
                    timestamp: new Date().toISOString()
                });

                const options = {
                    hostname: 'httpbin.org',
                    port: 443,
                    path: '/post',
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json',
                        'Content-Length': Buffer.byteLength(postData)
                    },
                    timeout: 10000
                };

                const req = https.request(options, (res) => {
                    const cipher = res.socket.getCipher();
                    console.log(`  Status: ${res.statusCode}`);
                    console.log(`  Cipher: ${cipher?.name}`);

                    let data = '';
                    res.on('data', (chunk) => { data += chunk; });
                    res.on('end', () => {
                        if (res.statusCode === 200) {
                            this.logTest('2.8', 'HTTPS POST Request', 'pass',
                                `POST request successful with ${cipher?.name}`);
                        } else {
                            this.logTest('2.8', 'HTTPS POST Request', 'fail',
                                `Unexpected status: ${res.statusCode}`);
                        }
                        resolve();
                    });
                });

                req.on('error', (error) => {
                    this.logTest('2.8', 'HTTPS POST Request', 'fail', error.message);
                    resolve();
                });

                req.on('timeout', () => {
                    req.destroy();
                    this.logTest('2.8', 'HTTPS POST Request', 'fail', 'Request timeout');
                    resolve();
                });

                req.write(postData);
                req.end();
            } catch (error) {
                this.logTest('2.8', 'HTTPS POST Request', 'fail', error.message);
                resolve();
            }
        });
    }

    async runAllTests() {
        console.log('='.repeat(60));
        console.log('Connectivity Tests');
        console.log('='.repeat(60));
        console.log('');

        await this.test_2_1_https_get_request();
        await this.test_2_2_tls_protocol_support();
        await this.test_2_3_tls_1_2_connection();
        await this.test_2_4_tls_1_3_connection();
        await this.test_2_5_certificate_validation();
        await this.test_2_6_cipher_suite_negotiation();
        await this.test_2_7_concurrent_connections();
        await this.test_2_8_https_post_request();

        console.log('='.repeat(60));
        console.log('Test Summary');
        console.log('='.repeat(60));
        console.log(`Total Tests: ${this.results.total_tests}`);
        console.log(`Passed: ${this.results.passed}`);
        console.log(`Failed: ${this.results.failed}`);
        console.log(`Skipped: ${this.results.skipped}`);

        const passRate = (this.results.passed / this.results.total_tests * 100).toFixed(1);
        console.log(`Pass Rate: ${passRate}%`);
        console.log('');

        if (this.results.passed >= 7) {
            console.log('✅ CONNECTIVITY TESTS PASSED');
            return 0;
        } else if (this.results.passed >= 5) {
            console.log('⚠️  PARTIAL SUCCESS (5-6/8 tests passed)');
            return 1;
        } else {
            console.log('❌ CRITICAL FAILURE (< 5/8 tests passed)');
            return 2;
        }
    }

    saveResults(filename = 'results.json') {
        try {
            fs.writeFileSync(filename, JSON.stringify(this.results, null, 2));
            console.log(`Results saved to: ${filename}`);
        } catch (error) {
            console.error(`Failed to save results: ${error.message}`);
        }
    }
}

// Run the tests
(async () => {
    const tests = new ConnectivityTests();
    const exitCode = await tests.runAllTests();
    tests.saveResults();
    process.exit(exitCode);
})();

File: python/3.13.7-bookworm-slim-fips/diagnostics/test-connectivity.py
#!/usr/bin/env python3
"""
Connectivity Tests
Tests basic SSL/TLS connectivity to real-world servers
"""

import ssl
import sys
import json
import os
import socket
import urllib.request
import urllib.error
from datetime import datetime, timezone
from concurrent.futures import ThreadPoolExecutor, as_completed
import time

class ConnectivityTests:
    def __init__(self):
        self.results = {
            "test_area": "2-connectivity",
            "timestamp": datetime.now(timezone.utc).isoformat().replace('+00:00', 'Z'),
            "container": "root-io-python-osp:latest",
            "total_tests": 8,
            "passed": 0,
            "failed": 0,
            "skipped": 0,
            "tests": []
        }
    
    def log_test(self, test_id, name, status, details="", duration_ms=0):
        """Log test result"""
        test_result = {
            "id": test_id,
            "name": name,
            "status": status,
            "duration_ms": duration_ms,
            "details": details
        }
        self.results["tests"].append(test_result)
        
        if status == "pass":
            self.results["passed"] += 1
            print(f"✅ {test_id} {name}: PASS")
        elif status == "fail":
            self.results["failed"] += 1
            print(f"❌ {test_id} {name}: FAIL")
        else:
            self.results["skipped"] += 1
            print(f"⏭️  {test_id} {name}: SKIP")
        
        if details:
            print(f"   Details: {details}")
        print()
    
    def https_get(self, url, timeout=10):
        """Perform HTTPS GET request"""
        try:
            # Create SSL context with explicit CA bundle
            context = ssl.create_default_context()
            context.load_verify_locations(cafile="/etc/ssl/certs/ca-certificates.crt")
            
            start = time.time()
            with urllib.request.urlopen(url, timeout=timeout, context=context) as response:
                status = response.status
                content_length = len(response.read())
                duration = int((time.time() - start) * 1000)
                return {
                    "success": True,
                    "status": status,
                    "content_length": content_length,
                    "duration_ms": duration
                }
        except urllib.error.HTTPError as e:
            return {
                "success": e.code < 500,  # 4xx is still a successful connection
                "status": e.code,
                "error": str(e),
                "duration_ms": 0
            }
        except Exception as e:
            return {
                "success": False,
                "error": str(e),
                "duration_ms": 0
            }
    
    def test_2_1_https_google(self):
        """Test 2.1: HTTPS GET - Google"""
        try:
            result = self.https_get("https://www.google.com")
            
            if result["success"] and 200 <= result.get("status", 0) < 400:
                details = f"Status: {result['status']}, Size: {result['content_length']} bytes, Time: {result['duration_ms']}ms"
                self.log_test("2.1", "HTTPS GET - Google", "pass", details, result['duration_ms'])
            else:
                error = result.get("error", f"Status: {result.get('status', 'unknown')}")
                self.log_test("2.1", "HTTPS GET - Google", "fail", f"Error: {error}")
        except Exception as e:
            self.log_test("2.1", "HTTPS GET - Google", "fail", f"Exception: {str(e)}")
    
    def test_2_2_https_github(self):
        """Test 2.2: HTTPS GET - GitHub"""
        try:
            result = self.https_get("https://www.github.com")
            
            if result["success"] and 200 <= result.get("status", 0) < 400:
                details = f"Status: {result['status']}, Size: {result['content_length']} bytes, Time: {result['duration_ms']}ms"
                self.log_test("2.2", "HTTPS GET - GitHub", "pass", details, result['duration_ms'])
            else:
                error = result.get("error", f"Status: {result.get('status', 'unknown')}")
                self.log_test("2.2", "HTTPS GET - GitHub", "fail", f"Error: {error}")
        except Exception as e:
            self.log_test("2.2", "HTTPS GET - GitHub", "fail", f"Exception: {str(e)}")
    
    def test_2_3_https_python(self):
        """Test 2.3: HTTPS GET - Python.org"""
        try:
            result = self.https_get("https://www.python.org")
            
            if result["success"] and 200 <= result.get("status", 0) < 400:
                details = f"Status: {result['status']}, Size: {result['content_length']} bytes, Time: {result['duration_ms']}ms"
                self.log_test("2.3", "HTTPS GET - Python.org", "pass", details, result['duration_ms'])
            else:
                error = result.get("error", f"Status: {result.get('status', 'unknown')}")
                self.log_test("2.3", "HTTPS GET - Python.org", "fail", f"Error: {error}")
        except Exception as e:
            self.log_test("2.3", "HTTPS GET - Python.org", "fail", f"Exception: {str(e)}")
    
    def test_2_4_https_api(self):
        """Test 2.4: HTTPS GET - API Endpoint"""
        try:
            result = self.https_get("https://api.github.com")
            
            if result["success"] and 200 <= result.get("status", 0) < 400:
                details = f"Status: {result['status']}, JSON API response, Time: {result['duration_ms']}ms"
                self.log_test("2.4", "HTTPS GET - API Endpoint", "pass", details, result['duration_ms'])
            else:
                error = result.get("error", f"Status: {result.get('status', 'unknown')}")
                self.log_test("2.4", "HTTPS GET - API Endpoint", "fail", f"Error: {error}")
        except Exception as e:
            self.log_test("2.4", "HTTPS GET - API Endpoint", "fail", f"Exception: {str(e)}")
    
    def test_2_5_tls_1_2_connection(self):
        """Test 2.5: TLS 1.2 Connection"""
        try:
            hostname = "www.google.com"
            port = 443
            
            # Create TLS 1.2 context
            context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
            context.minimum_version = ssl.TLSVersion.TLSv1_2
            context.maximum_version = ssl.TLSVersion.TLSv1_2
            context.check_hostname = True
            context.verify_mode = ssl.CERT_REQUIRED
            context.load_verify_locations(cafile="/etc/ssl/certs/ca-certificates.crt")
            
            start = time.time()
            with socket.create_connection((hostname, port), timeout=10) as sock:
                with context.wrap_socket(sock, server_hostname=hostname) as ssock:
                    version = ssock.version()
                    cipher = ssock.cipher()
                    duration = int((time.time() - start) * 1000)
                    
                    print(f"  TLS Version: {version}")
                    print(f"  Cipher: {cipher[0] if cipher else 'unknown'}")
                    
                    if version and "1.2" in version:
                        details = f"TLS {version}, Cipher: {cipher[0]}, Time: {duration}ms"
                        self.log_test("2.5", "TLS 1.2 Connection", "pass", details, duration)
                    else:
                        self.log_test("2.5", "TLS 1.2 Connection", "fail",
                                    f"Expected TLS 1.2, got: {version}")
        except Exception as e:
            self.log_test("2.5", "TLS 1.2 Connection", "fail", f"Exception: {str(e)}")
    
    def test_2_6_tls_1_3_connection(self):
        """Test 2.6: TLS 1.3 Connection"""
        try:
            hostname = "www.cloudflare.com"  # Known to support TLS 1.3
            port = 443
            
            # Create TLS 1.3 context
            context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
            context.minimum_version = ssl.TLSVersion.TLSv1_3
            context.check_hostname = True
            context.verify_mode = ssl.CERT_REQUIRED
            context.load_verify_locations(cafile="/etc/ssl/certs/ca-certificates.crt")
            
            start = time.time()
            with socket.create_connection((hostname, port), timeout=10) as sock:
                with context.wrap_socket(sock, server_hostname=hostname) as ssock:
                    version = ssock.version()
                    cipher = ssock.cipher()
                    duration = int((time.time() - start) * 1000)
                    
                    print(f"  TLS Version: {version}")
                    print(f"  Cipher: {cipher[0] if cipher else 'unknown'}")
                    
                    if version and "1.3" in version:
                        details = f"TLS {version}, Cipher: {cipher[0]}, Time: {duration}ms"
                        self.log_test("2.6", "TLS 1.3 Connection", "pass", details, duration)
                    else:
                        self.log_test("2.6", "TLS 1.3 Connection", "fail",
                                    f"Expected TLS 1.3, got: {version}")
        except Exception as e:
            # TLS 1.3 might not be supported everywhere, so log but don't fail hard
            self.log_test("2.6", "TLS 1.3 Connection", "fail", f"Exception: {str(e)}")
    
    def test_2_7_certificate_chain(self):
        """Test 2.7: Certificate Chain Validation"""
        try:
            hostname = "www.github.com"
            port = 443
            
            context = ssl.create_default_context()
            context.load_verify_locations(cafile="/etc/ssl/certs/ca-certificates.crt")
            
            start = time.time()
            with socket.create_connection((hostname, port), timeout=10) as sock:
                with context.wrap_socket(sock, server_hostname=hostname) as ssock:
                    cert = ssock.getpeercert()
                    duration = int((time.time() - start) * 1000)
                    
                    # Extract certificate details
                    subject = dict(x[0] for x in cert.get('subject', []))
                    issuer = dict(x[0] for x in cert.get('issuer', []))
                    
                    print(f"  Subject: {subject.get('commonName', 'unknown')}")
                    print(f"  Issuer: {issuer.get('commonName', 'unknown')}")
                    print(f"  Valid from: {cert.get('notBefore', 'unknown')}")
                    print(f"  Valid until: {cert.get('notAfter', 'unknown')}")
                    
                    if cert and subject:
                        details = f"Valid cert chain, CN: {subject.get('commonName')}, Time: {duration}ms"
                        self.log_test("2.7", "Certificate Chain Validation", "pass", details, duration)
                    else:
                        self.log_test("2.7", "Certificate Chain Validation", "fail",
                                    "No certificate retrieved")
        except Exception as e:
            self.log_test("2.7", "Certificate Chain Validation", "fail", f"Exception: {str(e)}")
    
    def test_2_8_concurrent_connections(self):
        """Test 2.8: Concurrent Connections"""
        try:
            urls = [
                "https://www.google.com",
                "https://www.github.com",
                "https://www.python.org",
                "https://www.cloudflare.com",
                "https://www.microsoft.com",
                "https://www.amazon.com",
                "https://www.apple.com",
                "https://www.netflix.com",
                "https://www.wikipedia.org",
                "https://www.reddit.com"
            ]
            
            successful = 0
            failed = 0
            results_list = []
            
            start = time.time()
            with ThreadPoolExecutor(max_workers=10) as executor:
                future_to_url = {executor.submit(self.https_get, url, 15): url for url in urls}
                
                for future in as_completed(future_to_url):
                    url = future_to_url[future]
                    try:
                        result = future.result()
                        if result["success"]:
                            successful += 1
                        else:
                            failed += 1
                        results_list.append((url, result["success"]))
                    except Exception as e:
                        failed += 1
                        results_list.append((url, False))
            
            total_duration = int((time.time() - start) * 1000)
            
            print(f"  Successful connections: {successful}/{len(urls)}")
            print(f"  Failed connections: {failed}/{len(urls)}")
            
            if successful >= 9:  # 9/10 or better
                details = f"{successful}/{len(urls)} connections successful, Total time: {total_duration}ms"
                self.log_test("2.8", "Concurrent Connections", "pass", details, total_duration)
            elif successful >= 7:  # 7-8/10
                details = f"{successful}/{len(urls)} connections successful (acceptable), Total time: {total_duration}ms"
                self.log_test("2.8", "Concurrent Connections", "pass", details, total_duration)
            else:
                details = f"Only {successful}/{len(urls)} connections successful"
                self.log_test("2.8", "Concurrent Connections", "fail", details)
        except Exception as e:
            self.log_test("2.8", "Concurrent Connections", "fail", f"Exception: {str(e)}")
    
    def run_all_tests(self):
        """Run all connectivity tests"""
        print("=" * 60)
        print("Connectivity Tests")
        print("=" * 60)
        print()
        
        self.test_2_1_https_google()
        self.test_2_2_https_github()
        self.test_2_3_https_python()
        self.test_2_4_https_api()
        self.test_2_5_tls_1_2_connection()
        self.test_2_6_tls_1_3_connection()
        self.test_2_7_certificate_chain()
        self.test_2_8_concurrent_connections()
        
        print("=" * 60)
        print("Test Summary")
        print("=" * 60)
        print(f"Total Tests: {self.results['total_tests']}")
        print(f"Passed: {self.results['passed']}")
        print(f"Failed: {self.results['failed']}")
        print(f"Skipped: {self.results['skipped']}")
        
        pass_rate = (self.results['passed'] / self.results['total_tests']) * 100
        print(f"Pass Rate: {pass_rate:.1f}%")
        print()
        
        if self.results['passed'] >= 7:
            print("✅ CONNECTIVITY TESTS PASSED")
            return 0
        elif self.results['passed'] >= 5:
            print("⚠️  PARTIAL SUCCESS (5-6/8 tests passed)")
            return 1
        else:
            print("❌ CRITICAL FAILURE (< 5/8 tests passed)")
            return 2
    
    def save_results(self, filename="results.json"):
        """Save results to JSON file"""
        output_path = os.path.join(os.path.dirname(__file__), filename)
        with open(output_path, 'w') as f:
            json.dump(self.results, f, indent=2)
        print(f"Results saved to: {output_path}")

if __name__ == "__main__":
    tests = ConnectivityTests()
    exit_code = tests.run_all_tests()
    tests.save_results()
    sys.exit(exit_code)

